hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/metadata/HazelcastRelMdPrunability.java

```diff
@@ -0,0 +1,185 @@
+import com.hazelcast.jet.sql.impl.opt.metadata.HazelcastRelMdPrunability.PrunabilityMetadata;
+import com.hazelcast.jet.sql.impl.opt.physical.FullScanPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.IndexScanMapPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.prunability.PartitionStrategyConditionExtractor;
+import com.hazelcast.jet.sql.impl.schema.HazelcastTable;
+import com.hazelcast.sql.impl.extract.QueryPath;
+import com.hazelcast.sql.impl.schema.TableField;
+import com.hazelcast.sql.impl.schema.map.PartitionedMapTable;
+import org.apache.calcite.linq4j.tree.Types;
+import org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.calcite.rel.BiRel;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.Calc;
+import org.apache.calcite.rel.metadata.Metadata;
+import org.apache.calcite.rel.metadata.MetadataDef;
+import org.apache.calcite.rel.metadata.MetadataHandler;
+import org.apache.calcite.rel.metadata.ReflectiveRelMetadataProvider;
+import org.apache.calcite.rel.metadata.RelMetadataProvider;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.Util;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyMap;
+
+public final class HazelcastRelMdPrunability
+        implements MetadataHandler<PrunabilityMetadata> {
+
+    public static final RelMetadataProvider SOURCE = ReflectiveRelMetadataProvider.reflectiveSource(
+            PrunabilityMetadata.METHOD,
+            new HazelcastRelMdPrunability()
+    );
+
+    private HazelcastRelMdPrunability() {
+    }
+
+    @Override
+    public MetadataDef<PrunabilityMetadata> getDef() {
+        return PrunabilityMetadata.DEF;
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(
+            FullScanPhysicalRel scan,
+            RelMetadataQuery mq
+    ) {
+
+        final HazelcastTable hazelcastTable = OptUtils.extractHazelcastTable(scan);
+        if (!(hazelcastTable.getTarget() instanceof PartitionedMapTable)) {
+            return emptyMap();
+        }
+
+        final PartitionedMapTable targetTable = hazelcastTable.getTarget();
+        final HashSet<String> partitioningFiledNames = new HashSet<>(targetTable.partitioningAttributes());
+        final Set<String> partitioningColumns = targetTable.keyFields()
+                .filter(kf -> partitioningFiledNames.contains(kf.getPath().getPath()))
+                .map(TableField::getName)
+                .collect(Collectors.toSet());
+
+        if (targetTable.partitioningAttributes().isEmpty()) {
```

Comment:
@ivanthescientist it seems that `PartitionedMapTable.partitioningAttributes()` does not contain `__key` but is empty. If that is the convention it should be documented because it is not obvious, as `__key` technically is partitioning key when there is no AttributePartitioningStrategy.